File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let lp = 0;
    let rp = height.length-1;
    let lm = 0;
    let rm =0;
    let twl = 0;
    while(lp < rp){
        if(height[lp] < height[rp]){
            lm = Math.max(lm,height[lp]);
            twl += lm- height[lp];
            lp++;
        }else{
            rm = Math.max(rm,height[rp]);
            twl += rm- height[rp];
            rp--;
        }
    }
    return twl;
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        Stack<Integer> st = new Stack<>();
        int res = 0;
        for(int i = 0; i < height.length;i++){
            while(!st.isEmpty() && height[st.peek()] < height[i]){
                int pop_height = height[st.pop()];
                if(st.isEmpty()){
                    break;
                }
                int dis = i-st.peek()-1;
                int min_height = Math.min(height[i],height[st.peek()])-pop_height;
                res += min_height*dis;
            }
            st.push(i);
        }
    return res;
    }
}